src/librustdoc/html/render/mod.rs

```diff
@@ -737,42 +737,73 @@ fn render_impls(
     w.write_str(&rendered_impls.join(""));
 }
 
-fn naive_assoc_href(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
-    use crate::formats::item_type::ItemType::*;
+/// Build a possibly empty `href` attribute (a key-value pair) for the given associated item.
+fn assoc_href_attr(it: &clean::Item, link: AssocItemLink<'_>, cx: &Context<'_>) -> String {
+    let name = it.name.unwrap();
+    let item_type = it.type_();
 
-    let name = it.name.as_ref().unwrap();
-    let ty = match it.type_() {
-        Typedef | AssocType => AssocType,
-        s => s,
-    };
+    let href = match link {
+        AssocItemLink::Anchor(Some(ref id)) => Some(format!("#{}", id)),
+        AssocItemLink::Anchor(None) => Some(format!("#{}.{}", item_type, name)),
+        AssocItemLink::GotoSource(did, provided_items) => {
+            use ItemType::*;
+
+            let (ty_provided_item, ty_required_item, kind) = match item_type {
+                Method | TyMethod => (Method, TyMethod, ty::AssocKind::Fn),
+                AssocConst | TyAssocConst => (AssocConst, TyAssocConst, ty::AssocKind::Const),
+                AssocType | TyAssocType => (AssocType, TyAssocType, ty::AssocKind::Type),
+                _ => panic!("assoc_href_attr called on non-associated item"),
+            };
 
-    let anchor = format!("#{}.{}", ty, name);
-    match link {
-        AssocItemLink::Anchor(Some(ref id)) => format!("#{}", id),
-        AssocItemLink::Anchor(None) => anchor,
-        AssocItemLink::GotoSource(did, _) => {
-            href(did.expect_def_id(), cx).map(|p| format!("{}{}", p.0, anchor)).unwrap_or(anchor)
+            // We're creating a link from an impl-item to the corresponding
+            // trait-item and need to map the anchored type accordingly.
+            let provided = provided_items.contains(&(name, kind));
+            let ty = if provided { ty_provided_item } else { ty_required_item };
+
+            match href(did.expect_def_id(), cx) {
+                Ok((url, ..)) => Some(format!("{}#{}.{}", url, ty, name)),
+                // FIXME: does this case still occur? should we panic instead?
```

Comment:
Thanks for adding it!

Reply:
Now since I came into contact with cross-crate documentation issues (#95717), I think this code path (`DocumentationNotBuilt`) is reached when someone implements a trait from an external crate whose documentation wasn't built. I don't know if that ever happens with `cargo doc` but it sure happens with plain `rustdoc`:

Let there be crate `a` defining `pub trait Tr { fn f(); }` and crate `b` depending on `a` containing `pub struct Ty; impl a::Tr for Ty { fn f() {} }`.
Now execute

```sh
rustc a.rs --crate-type=lib --crate-name=a --edition=2021
rustdoc b.rs --crate-type=lib --crate-name=b --edition=2021 --extern=a=liba.rlib
```

In such case, the `f` in the documentation of `impl a::Tr for Ty` does **not** have an `href` attribute since the link would be broken.

Reply:
Now since I came into contact with cross-crate documentation issues (#95717), I think this code path (`DocumentationNotBuilt`) is reached when someone implements a trait from an external crate whose documentation wasn't built. I don't know if that ever happens with `cargo doc` but it sure happens with plain `rustdoc`:

Let there be crate `a` defining `pub trait Tr { fn f(); }` and crate `b` depending on `a` containing `pub struct Ty; impl a::Tr for Ty { fn f() {} }`.
Now execute

```sh
rustc a.rs --crate-type=lib --crate-name=a --edition=2021
rustdoc b.rs --crate-type=lib --crate-name=b --edition=2021 --extern=a=liba.rlib
```

In such case, the `f` in the documentation of `impl a::Tr for Ty` does **not** have an `href` attribute since the link would be broken.